lib/issuesyncgithub/github.go

```diff
@@ -12,39 +13,93 @@ import (
 	"golang.org/x/oauth2"
 )
 
-// GitHubClient is a wrapper around the GitHub API Client library we
+// Client is a wrapper around the GitHub API Client library we
 // use. It allows us to swap in other implementations, such as a dry run
 // clients, or mock clients for testing.
-type GitHubClient interface {
-	ListIssues() ([]github.Issue, error)
+type Client interface {
+	ListIssues() ([]models.ExtendedGithubIssue, error)
+	//GetCurrentProjectCardAndCommitIds(issue *github.Issue) (*github.ProjectCard, error)
 	ListComments(issue github.Issue) ([]*github.IssueComment, error)
 	GetUser(login string) (github.User, error)
 	GetRateLimits() (github.RateLimits, error)
 }
 
 // realGHClient is a standard GitHub clients, that actually makes all of the
 // requests against the GitHub REST API. It is the canonical implementation
-// of GitHubClient.
+// of Client.
 type realGHClient struct {
 	config cfg.Config
 	client github.Client
 }
 
-// ListIssues returns the list of GitHub issues since the last run of the tool.
-func (g realGHClient) ListIssues() ([]github.Issue, error) {
+func (g realGHClient) GetCurrentProjectCardAndCommitIds(issue *github.Issue) (*github.ProjectCard, []string, error) {
 	log := g.config.GetLogger()
 
 	ctx := context.Background()
 
 	user, repo := g.config.GetRepo()
 
+	pages := 1
+
+	var currentProjectCard *github.ProjectCard
+
+	var commitIds []string
+
+	// search for the current project card
+	for page := 1; page <= pages; page++ {
+		is, res, err := g.request(func() (interface{}, *github.Response, error) {
+			return g.client.Issues.ListIssueEvents(ctx, user, repo, issue.GetNumber(), &github.ListOptions{
+				Page:    page,
+				PerPage: 100,
+			})
+		})
+		if err != nil {
+			return nil, nil, err
+		}
+		issueEventPointers, ok := is.([]*github.IssueEvent)
+		if !ok {
+			log.Errorf("Get GitHub issue events did not return issue events! Got: %v", is)
+			return nil, nil, fmt.Errorf("get GitHub issues events failed: expected []*github.IssueEvent; got %T", is)
+		}
+
+		for _, v := range issueEventPointers {
+			if v.ProjectCard != nil {
+				if v.GetEvent() == "added_to_project" || v.GetEvent() == "moved_columns_in_project" {
+					currentProjectCard = v.ProjectCard
+				} else if *v.Event == "removed_from_project" {
+					currentProjectCard = nil
+				}
+			}
+
+			if v.CommitID != nil {
+				commitIds = append(commitIds, *v.CommitID)
+			}
+		}
+		pages = res.LastPage
+	}
+
+	log.Debug("Found current Project Card for issue #%d", issue.GetNumber())
+
+	return currentProjectCard, commitIds, nil
+}
+
+// ListIssues returns the list of GitHub issues since the last run of the tool.
+func (g realGHClient) ListIssues() ([]models.ExtendedGithubIssue, error) {
+	log := g.config.GetLogger()
+
+	ctx := context.Background()
```

Comment:
Small thing, but I would expect ctx to be the context, rather than the background.

Reply:
Context is the package. Background is the specific context being pulled out of the context package.

Reply:
Yuck. Then I would favor a name that exposes this bit of go weirdness, like: backgroundContext := context.Background().